src/LibUsbDotNet/LibUsb/AsyncTransfer.cs

```diff
@@ -19,80 +19,94 @@
-    private static IntPtr transferDelegatePtr = Marshal.GetFunctionPointerForDelegate(transferDelegate);
+    private static readonly unsafe TransferDelegate TransferCallback = new TransferDelegate(Callback);
+    private static readonly IntPtr TransferDelegatePtr = 
+        Marshal.GetFunctionPointerForDelegate(TransferCallback);
+    private static readonly ConcurrentDictionary<int, TaskCompletionSource<(Error error, int transferLength)>>
+        TransferDictionary = new ConcurrentDictionary<int, TaskCompletionSource<(Error error, int transferLength)>>();
 
-    public static Error TransferAsync(
+    public static Task<(Error error, int transferLength)> TransferAsync(
         DeviceHandle device,
         byte endPoint,
         EndpointType endPointType,
         IntPtr buffer,
         int offset,
         int length,
-        int timeout,
-        out int transferLength)
-    {
-        return TransferAsync(device, endPoint, endPointType, buffer, offset, length, timeout, 0, out transferLength);
-    }
+        int timeout) =>
+        TransferAsync(device, endPoint, endPointType, buffer, offset, length, timeout, 0);
 
-    internal static unsafe Error TransferAsync(
+    public static async Task<(Error error, int transferLength)> TransferAsync(
         DeviceHandle device,
         byte endPoint,
         EndpointType endPointType,
         IntPtr buffer,
         int offset,
         int length,
         int timeout,
-        int isoPacketSize,
-        out int transferLength)
+        int isoPacketSize)
     {
         if (device == null)
-        {
             throw new ArgumentNullException(nameof(device));
-        }
 
         if (offset < 0)
-        {
             throw new ArgumentOutOfRangeException(nameof(offset));
-        }
-
-        // Determin the amount of isosynchronous packets
-        int numIsoPackets = 0;
 
-        if (isoPacketSize > 0)
+        var transferCompletion =
+            new TaskCompletionSource<(Error error, int transferLength)>(TaskCreationOptions.RunContinuationsAsynchronously);
+            
+        int transferId;
+        lock (TransferLock)
         {
-            numIsoPackets = length / isoPacketSize;
+            if (_transferIndex == int.MaxValue) // Potential edge case for long-running application?
+                _transferIndex = 0;
+            transferId = _transferIndex++;
         }
 
-        var transfer = NativeMethods.AllocTransfer(numIsoPackets);
+        if (!TransferDictionary.TryAdd(transferId, transferCompletion))
+            throw new InvalidOperationException(
+                $"{transferId} already exists in {nameof(TransferDictionary)}");
 
-        ManualResetEventSlim mre = new ManualResetEventSlim(false);
+        FillAndSubmitTransfer(device, endPoint, endPointType, buffer, offset, length, timeout, isoPacketSize,
+            transferId);
 
-        int transferId = 0;
+        return await transferCompletion.Task.ConfigureAwait(false);
```

Comment:
nit: you can slightly simplify this code by making the method return `transferCompletion.Task` without awaiting it, e.g.:

```csharp
    public static Task<(Error error, int transferLength)> TransferAsync(
        DeviceHandle device,
        byte endPoint,
        EndpointType endPointType,
        IntPtr buffer,
        int offset,
        int length,
        int timeout,
        int isoPacketSize)
{
    ...

    return transferCompletion.Task;
}
```